Added docs section for starting contributors

Jwaegebaert opened the pull request:
Finally, after a few months of work, lots of head-scratching, I can deliver the first design for a new section in the docs. A guide on how to get started with contributing to the CLI for Microsoft 365. 

The following topics are included in this section.

- Contributing guide
- Environment setup
- Creating a new command: 
  - New command sample
  - Command logic
  - Unit tests
  - Help page
- Adding a script sample
- Debugging the CLI
- Creating a Pull Request
- Pull Request reviews

@pnp/cli-for-microsoft-365-maintainers a thorough review would be nice 😄 

---

Closes #2834

Adam-it commented:
Awesome work @Jwaegebaert . We will review it ASAP 🤩

waldekmastykarz commented:
@garrytrinder, I believe you had some ideas in this space. Would you mind reviewing this PR?

garrytrinder commented:
Great work @Jwaegebaert 👏

I've had a quick look and here are some of my initial thoughts.

- Consider changing the `Contributing` landing page to provide an overview and index of the pages under this section, the current page doesn't help guide me to my next step. An index helps provide an end to end view of the guide as you can see all the steps within that guide.
- Consider providing two routes to tutorials on the landing page, one for contributing a new command and the other a sample script. Each route has a potentially different audience, some folks just want to contribute scripts, so may focus on this first.
- Consider using a different command name or example that doesn't currently exist i.e. `foo` command. The `group get` command already exists, so should we consider using something different to avoid conflicts.
- Consider including numbers in the headings to make it clear that the steps are to be followed in order.
- Consider that the guide should focus on instructing the reader to perform actions and see immediate results to those actions. We should be very much handholding them through the process for creating a command or a script. Currently I don't feel that we do this and leave things to interpretation. Let's make sure they are setup and ready like having `npm watch` running.

garrytrinder commented:
Apologies on the delay on getting back to you on this, life has been getting in the way.

> Do you maybe know how we can achieve this with `Material for MkDocs`? I'm asking this because I had the same thought but didn't find a proper way of visually implementing this.

This could just be a couple of [buttons](https://squidfunk.github.io/mkdocs-material/reference/buttons/), or even simpler just a bulleted list.

> This was my initial thought also, but won't it be easier for users to have a fully implemented example? That way they can always validate the existing files in the project and reflect on this.

It might be easier, but as the tutorial is aimed at folks with no prior knowledge, we could guide them through setting up the development environment, starting watch mode, build and run a command and then write the tests. This would show the contributor the full end to end process and at the end they have built something from scratch. This would be a great way to learn. It's also a good way to show the developer experience.

Jwaegebaert commented:
Alright, it has been a while since I last updated this PR. Here are the changes I have made:

- Added a "Next Steps" section at the end of several articles to guide users toward follow-up articles.
- Included a few pathways on the landing page that users can follow to achieve specific goals.
- Updated some content to improve readability and ease of following.

I want to mention that I haven't incorporated the suggestions for the new command example. Personally, I believe it is much easier for users to understand and follow a real-world scenario rather than a hypothetical one. This way, users can also refer to the fully implemented files to compare their own scenarios.

martinlingstuyl commented:
Do you need help on this one @garrytrinder? It would be nice if we can merge it before the migration to docusaurus.

martinlingstuyl commented:
@Jwaegebaert, @garrytrinder: just to get the lay of the land: how far along is this PR? Do you think it's just crossing some t's and dotting some I's? Or might more zealous reading be in order?

Jwaegebaert commented:
Some zealous reading is always welcome 😄 

I've done a second round of adjustments so it could always do with some tweaks here and there. But with Docusaurus around the corner like you said. We'll need to make some adjustments to this guide to include the Docusaurus architecture

martinlingstuyl commented:
In that case maybe we should leave it until after the docusaurus release...

milanholemans reviewed (commented):
We've recently changed our SSG to Docusaurus. This update has impacted our documentation syntaxes and will require some adjustments on your end. However, there's no need to worry, as Milan has done an amazing job creating a detailed page that explains all the new syntaxes to be used in the .mdx files. If you have any questions, feel free to reach out to us here or on Discord. You can find more information at this link: [All you need to know about Docusaurus](https://github.com/pnp/cli-microsoft365/wiki/All-you-have-to-know-about-Docusaurus).

😉

Jwaegebaert commented:
Closing this to test Parker

garrytrinder commented:
> @Jwaegebaert, @garrytrinder: just to get the lay of the land: how far along is this PR? Do you think it's just crossing some t's and dotting some I's? Or might more zealous reading be in order?

This had fallen off my list, however now that Docusaurus release has happened it seems like a good time to revisit this.

Jwaegebaert commented:
It's been updated to include all the changes regarding Docusaurus

garrytrinder commented on docs/docs/contribute/environment-setup.mdx:
Should we also include alternative approaches here, such as using Remote Development containers or GitHub Codespaces? These approaches handle a lot of the manual steps mentioned in this page for contributors and can help remove the friction to getting started.

Jwaegebaert replied:
We'll ... I don't have any experience with remote dev 😅 So I'm not quite sure what to put here

garrytrinder replied:
OK, let's make sure we create an issue to add alternative approaches so we don't forget about it and hold up this PR, we can always iterate after.

garrytrinder commented on docs/docs/contribute/contributing-guide.mdx:
```diff
@@ -0,0 +1,57 @@
+
+Sharing is caring!
+
+## Introduction to CLI for Microsoft 365
```
Do we need this section? I would assume that if you are thinking about contributing to the CLI you will already know what it is so seems unnecessary to repeat it here.

garrytrinder commented on docs/docs/contribute/creating-the-pr.mdx:
```diff
@@ -0,0 +1,71 @@
+
+## Check in your Local Changes
+
+Before proceeding, ensure that all your files are committed to your branch. If you're using Visual Studio Code, you can achieve this by selecting `Source Control` from the activity bar and providing a commit name in the `Source Control` tab. Generally, the commit name should be the title of the issue you are working on. Then click the commit button to safely commit your files to your local branch ...
```
It would be great to provide a real world example of a commit message and the format we use to help provide more context.

garrytrinder commented on docs/docs/contribute/creating-the-pr.mdx:
````diff
@@ -0,0 +1,71 @@
+If you're solely working with git, follow these steps:
+
+```bash
+# Stash all your files
````
`git add .` stages all modified files, it doesn't `stash` them. 

Stashing is a different concept, let's not introduce confusion of the terms for first time folks.

garrytrinder commented on docs/docs/contribute/creating-the-pr.mdx:
```diff
@@ -0,0 +1,71 @@
+# Stash all your files
+git add .
+
+# Commit your stash
```
```suggestion
# Commit staged changes
```

garrytrinder commented on docs/docs/contribute/creating-the-pr.mdx:
```diff
@@ -0,0 +1,71 @@
+git add .
+
+# Commit your stash
+git commit -m 'Adds command spo group get'
```
I think we should include the `.Closes #n` part of our commit message here. So that we reinforce that our contributors should be working from an issue.

garrytrinder commented on docs/docs/contribute/environment-setup.mdx:
```diff
@@ -0,0 +1,82 @@
+
+:::tip
+
+If you have installed the CLI globally using the `npm i -g @pnp/cli-microsoft365` command, we recommend that you uninstall it before running `npm link`.
```
Let's provide the uninstall command as well so we can help provide the next step.

garrytrinder commented on docs/docs/contribute/creating-the-pr.mdx:
````diff
@@ -0,0 +1,71 @@
+## Push your Local Branch
+
+Now we need to get all our local changes in our forked repository. This way we can create a new PR using GitHub.
+```bash
````
Looks like we are missing a line return here in between the text and code block.

garrytrinder reviewed (commented):
Great work @Jwaegebaert! Please find my initial comments. I've not had chance to review all pages, so this review is ongoing, but wanted to provide some initial feedback for you to consider.

Jwaegebaert commented:
Awesome initial review @garrytrinder, I've already applied most of these.

garrytrinder commented on docs/docs/contribute/environment-setup.mdx:
```diff
@@ -0,0 +1,82 @@
+
+:::info "Install prerequisites"
+
+Before you start contributing to this project, you will need **Node.js@18** and **npm@8** installed.
```
Minor thing but we usually drop `.js` when referring to a specific version on node.

```suggestion
Before you start contributing to this project, you will need **node@18** and **npm@8** installed.
```

garrytrinder commented on docs/docs/contribute/environment-setup.mdx:
```diff
@@ -0,0 +1,82 @@
+
+Although you can use any IDE, we recommend using Visual Studio Code. When using VS Code, the following extensions will come in handy:
+
+Extension | Why is it useful?
```
We should also include `Test Explorer UI` extension in this list. The `Mocha Test Explorer` extension extends the functionality of the `Test Explorer UI` to support `mocha` tests.

https://marketplace.visualstudio.com/items?itemName=hbenl.vscode-test-explorer

garrytrinder commented on docs/docs/contribute/environment-setup.mdx:
```diff
@@ -0,0 +1,82 @@
+`npm run watch` | Builds the entire project first. After this, a watcher will trigger an incremental build every time a file is saved. This means that only the changed files are rebuilt, not the entire project. **You typically use this command while developing**.
+`npm run build` | Builds the entire project.
+`npm run clean` | Cleans the output directory by deleting all built files.
+`npm run test` | Runs all tests and checks all ESLint rules. This is a combination of `npm run test:cov` and `npm run lint`. This command is executed in our GitHub workflows when creating a PR.
+`npm run test:cov` | Runs all tests and creates a coverage report.
+<nobr>`npm run test:test`</nobr> | Runs all tests.
+`npm run lint` | Runs all ESLint rules.
```
As tests, linting rules and code coverage checks are all executed when you run `npm run test`, I think we can ignore the sub commands that are called. First time contributors really just need to know how to watch, build and run tests IMO.

```suggestion
`npm run test` | Executes all tests and linting rules, also generates a code coverage report. This command is executed automatically when you raise a pull request.
```

garrytrinder commented on docs/docs/contribute/expect-during-pr.mdx:
```diff
@@ -0,0 +1,37 @@
+
+When your PR is under review, the assigned maintainer will indicate their involvement by assigning themselves to the PR. They will follow a checklist for the review process, which can be found [here](https://github.com/pnp/cli-microsoft365/wiki/PR-checklist).
+
+Maintainers may also provide suggestions for best practices to ensure consistent coding styles. If there are code-related comments, you will see "changes requested" next to your username in the PR. In case the maintainer needs your input on an issue before continuing the review, the PR will be labeled as "waiting on response." The reviewer may also mark the pull request as **Draft** to indicate t ...
```
```suggestion
Maintainers may also provide suggestions for best practices to ensure consistent coding styles. If there are code-related comments, you will see "changes requested" next to your username in the PR. In case the maintainer needs your input on an issue before continuing the review, the PR will be labelled as "waiting on response." The reviewer will also mark the pull request as **Draft** to indicate that it is `work in progress` and that changes are expected to be made.
```

garrytrinder commented on docs/docs/contribute/how-to-debug.mdx:
```diff
@@ -0,0 +1,107 @@
+
+:::
+
+### Using the AddOn 'Test Explorer'
```
```suggestion
### Using Test Explorer UI VSCode Extension
```

garrytrinder reviewed (commented):
Thanks for the updates @Jwaegebaert here are some more comments 👍

milanholemans reviewed (changes requested):
Hi, @Jwaegebaert awesome work! A lot of valuable content!

I noticed this PR has been stale for a while so I checked your PR (because I find this topic interesting), and noticed some outdated info. I'm not trying to take over @garrytrinder 's role; he's doing great. Just wanted to share my observations and help improve the content.

Keep up the good work!

Jwaegebaert commented on docs/docs/contribute/contributing-guide.mdx:
```diff
@@ -0,0 +1,53 @@
+
+## You have an idea for a New Command
+
+Awesome! Good ideas are invaluable for every product. Before you start hacking away, please check if there is no similar idea already listed in the [issue list](https://github.com/pnp/cli-microsoft365/issues). This ensures that your idea is unique and not being worked on. If you don't find a similar idea, create a new issue describing your idea. Once we agree on the feature scope and architecture ...
```
Right now, there's no specific page for this, but it would be great to have one on the website. Maybe we can create a new issue to address this. We'll review the existing wiki pages and make sure to remove any links if we decide to move the content there.

milanholemans commented on docs/docs/contribute/how-to-debug.mdx:
````diff
@@ -0,0 +1,107 @@
+Translating this to the `args` array, it will look like this:
+
+```json
+args: ["m365", "spo", "group", "get", "--id", "1", "--webUrl", "https://contoso.sharepoint.com/sites/contoso"]
````
Noticed another thing.

```suggestion
args: ["spo", "group", "get", "--id", "1", "--webUrl", "https://contoso.sharepoint.com/sites/contoso"]
```

milanholemans commented on docs/docs/contribute/environment-setup.mdx:
```diff
@@ -0,0 +1,80 @@
+
+Extension | Why is it useful?
+--- | ---
+[Test Explorer UI](https://marketplace.visualstudio.com/items?itemName=hbenl.vscode-test-explorer) | This extension provides a graphical user interface for running tests and viewing the results. It also allows you to debug tests directly from the editor.
```
Not sure if we should since this extension is being installed automatically when installing `Test Explorer UI`.

Jwaegebaert commented on docs/src/config/sidebars.js:
Not really as the articles shouldn't be sorted alphabetically

milanholemans commented on docs/docs/contribute/environment-setup.mdx:
```diff
@@ -0,0 +1,80 @@
+
+:::
+
+## Visual Studio Code extensions
```
An extension we could add: https://marketplace.visualstudio.com/items?itemName=unifiedjs.vscode-mdx

Jwaegebaert commented:
Nice review @milanholemans, very thorough! Most suggestions are applied. One thing I was curious about, I noticed there weren't any suggestions regarding the file `contributing-guide.mdx`. Maybe you forgot that one?

milanholemans commented:
> Nice review @milanholemans, very thorough! Most suggestions are applied. One thing I was curious about, I noticed there weren't any suggestions regarding the file `contributing-guide.mdx`. Maybe you forgot that one?

I left 1 comment there 😃. From what I can remember, that page was good 👍

Jwaegebaert commented:
@pnp/cli-for-microsoft-365-maintainers can we give this PR a shot and merge it in? 😄

garrytrinder commented:
@Jwaegebaert LGTM 🚀

milanholemans commented:
Merged manually, nice work @Jwaegebaert!
